lib/shrinkwrap.js

```diff
@@ -0,0 +1,118 @@
+const path = require('path');
+
+const _ = require('lodash');
+const traverse = require('traverse');
+
+const deps = require('./util/deps');
+const exec = require('./util/exec');
+const link = require('./util/link');
+const pfs = require('./util/pfs');
+const promise = require('./util/promise');
+
+module.exports = function(dir) {
+  var dependencyKey = this.dependencyKey;
+  var moduleDir = this.moduleDir;
+  var log = this.emit.bind(this);
+
+  var shrinkwrapCache = {};
+  var shrinkwrapNodeDeps = _.memoize(function(repo) {
+    var repoPath = repo.welder.repoPath;
+    var shrinkwrapPath = path.join(repoPath, 'npm-shrinkwrap.json');
+    var repoDeps = repo[dependencyKey];
+    // Compare names in welder dependencies against npm dependencies. If the
+    // dependency only exist in welder deps that unlink it so npm shrinkwrap
+    // won't error when it encounters the extraneous dependency.
+    var repoNpmDeps = repo.dependencies;
+    return promise.resolve(_.pairs(repoDeps))
+      .map(function(pair) {
+        var depName = pair[0];
+        if (!repoNpmDeps || !repoNpmDeps[depName]) {
+          return pfs.unlink(path.join(repoPath, moduleDir, depName));
+        }
+      })
+      .then(function() {
+        return exec('npm shrinkwrap', { cwd: repoPath })
+      })
+      .then(function(result) {
+        log('debug', 'Shrinkwrapping node dependencies ' + repo.name + '...');
+        if (result.code === 0) {
+          return pfs.readFile(shrinkwrapPath, 'utf8')
+            .then(function(shrinkwrap) {
+              result.shrinkwrap = JSON.parse(shrinkwrap);
+              return result;
+            })
+            .finally(function() {
+              return pfs.unlink(shrinkwrapPath);
+            });
+        }
+        return result;
+      });
+  }, deps.repoId);
+
+  return this.graph(dir, { localVersion: true })
+    .bind(this)
+    .then(function(graph) {
+      var root = _.clone(graph);
+      root.nodes = deps.uniqRepos(deps.flattenGraph(graph));
+      root.nodes = root.nodes.map(function(node) {
+        node = _.clone(node);
+        delete node.nodes;
+        return node;
+      });
+      return root;
+    })
+    .then(function(graph) {
+      var nodes = graph.nodes.slice();
+      nodes.unshift(graph);
+
+      return promise.resolve(nodes)
+        .parallel(function(repo) {
+          return shrinkwrapNodeDeps(repo)
+            .tap(function(result) {
+              if (result.code === 0) {
+                repo.nodeDependencies = result.shrinkwrap.dependencies;
+              }
+            });
+        })
+        .then(_.partialRight(exec.warnIfAnyExitedNonZero, log))
+        .return(graph);
+    })
+    .then(function(graph) {
+      return traverse(graph).map(function(node) {
```

Comment:
I don't know. Feels weird. I'm not sure how it'd be reused. I'm splitting it outside of the this overall function for now.